Description: Fix sed on forward slashes
Author: Lance Lin <lq27267@gmail.com>
Forwarded: https://github.com/xmake-io/xmake/pull/4300
Date: 23 Aug 2023

--- a/configure
+++ b/configure
@@ -3665,7 +3665,7 @@
 _replace_configvar_value() {
     local name="${1}"
     local value="${2}"
-    _ret="s/\${${name}}/${value}/g"
+    _ret="s@\${${name}}@${value}@g"
 }
 
 # generate configfile for the given target
