diff --git a/core/src/xmake/os/fscase.c b/core/src/xmake/os/fscase.c
index 917a5dc75..00eeb338a 100644
--- a/core/src/xmake/os/fscase.c
+++ b/core/src/xmake/os/fscase.c
@@ -38,9 +38,13 @@ tb_int_t xm_os_fscase(lua_State* lua)
     // check
     tb_assert_and_check_return_val(lua, 0);
 
+#if ((TB_VERSION_MAJOR * 100) + (TB_VERSION_MINOR * 10) + TB_VERSION_ALTER) >= 174
     tb_char_t const* path = luaL_checkstring(lua, 1);
     tb_check_return_val(path, 0);
 
     lua_pushinteger(lua, (tb_int_t)tb_file_fscase(path));
+#else
+    lua_pushinteger(lua, -1);
+#endif
     return 1;
 }
